### PR TITLE
update causal batch for ct2 and fix nf4

### PR DESCRIPTION
**Pull request by michaelfeil**
Two issues with the current implemented features:

- `--quantize bitsandbytes-nf4` crashes, as `scipy` is not installed, which is required for `import bitsandbytes`. Beyond I bumped the huggingface-libaries, to be compatible with `bitsandbytes`.
- `--quantize ct2` stops parallel streams. Reason is that CTranslate2 does not offer an interface for KV Caching. In the previous implementation, KV Cache was set to `None`. As the `CausalBatch` requires concurrent requests are cached, the stream stops if parallel streaming users is >1.

Lastly, I updated the docker image. Its available to pull.

### CODE REVIEW

**Comment by michaelfeil** [comment]
@Atry ready for review!

File: server/text_generation_server/models/ct2_causal_lm.py [hunk 1]
```diff
@@ -51,6 +60,434 @@
 tracer = trace.get_tracer(__name__)
 
 
+@dataclass
```
Comment (atmiguel) [inline server/text_generation_server/models/ct2_causal_lm.py:63]: nit, it looks like we tend to prefer to require keyword args for dataclasses (i.e. `kw_only=true`)

File: server/text_generation_server/models/ct2_causal_lm.py [hunk 2]
```diff
@@ -51,6 +60,434 @@
...
+                # We need to offset the mapping for each batch by the cumulative batch size
+                for k, v in batch.requests_idx_mapping.items():
+                    requests_idx_mapping[k] = v + start_index
+
+            # Slicing end index for this batch
+            end_index = start_index + len(batch)
+
+            # We only concatenate batches that did at least one step
+            # if batch.past_key_values is None:
+            #     raise ValueError("only concatenate prefilled batches")
```
Comment (atmiguel) [inline server/text_generation_server/models/ct2_causal_lm.py:338]: would prefer to remove any commented out code, unless there's a compelling reason to keep it

File: server/text_generation_server/models/ct2_causal_lm.py [hunk 3]
```diff
@@ -51,6 +60,434 @@
...
+            # And ensure that we can update tensors in-place
+            # if type(batch.past_key_values[0]) == tuple:
+            #     batch.past_key_values = [
+            #         [t.view(len(batch), -1, *t.shape[-2:]) for t in layer]
+            #         for layer in batch.past_key_values
+            #     ]
+            # elif len(batch.past_key_values[0][0].shape) == 3:
+            #     for layer in batch.past_key_values:
+            #         for k, t in enumerate(layer):
+            #             layer[k] = t.view(len(batch), -1, *t.shape[-2:])
```
Comment (atmiguel) [inline server/text_generation_server/models/ct2_causal_lm.py:388]: would prefer to remove this

File: server/text_generation_server/models/ct2_causal_lm.py [hunk 4]
```diff
@@ -51,6 +60,434 @@
...
+        #         past_seq_len = batch.max_input_length - 1
+        #         padded_past_values[
+        #             start_index:end_index, :, -past_seq_len:, :
+        #         ] = past_values[:, :, -past_seq_len:, :]
+        #         del past_values
+
+        #         # Update values
+        #         start_index = end_index
+
+        #     past_key_values.append([padded_past_keys, padded_past_values])
```
Comment (atmiguel) [inline server/text_generation_server/models/ct2_causal_lm.py:466]: would prefer to remove this

**Comment by atmiguel** [comment]
> You think you can merge it?

Unfortunately, I don't have write access on this repo. Presumably @fmmoret does?

And tbh, I don't have enough context on TGI to properly review the code.

**Review by fmmoret — APPROVED** [review]

**Pull request merged**